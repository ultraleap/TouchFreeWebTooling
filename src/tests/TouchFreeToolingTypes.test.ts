import {
    BitmaskFlags,
    FlagUtilities,
    HandChirality,
    HandType,
    InputType,
    InteractionType,
} from '../TouchFreeToolingTypes';

// TODO: Generate enum values from types themselves?

const interactionTypes: InteractionType[] = [
    InteractionType.GRAB,
    InteractionType.HOVER,
    InteractionType.PUSH,
    InteractionType.TOUCHPLANE,
    InteractionType.VELOCITYSWIPE,
];
const handTypes: HandType[] = [HandType.PRIMARY, HandType.SECONDARY];
const handChiralities: HandChirality[] = [HandChirality.LEFT, HandChirality.RIGHT];
const inputTypes: InputType[] = [InputType.CANCEL, InputType.DOWN, InputType.MOVE, InputType.NONE, InputType.UP];

const allHandChiralityBitmask: BitmaskFlags = BitmaskFlags.LEFT | BitmaskFlags.RIGHT;
const allHandTypeBitmask: BitmaskFlags = BitmaskFlags.PRIMARY | BitmaskFlags.SECONDARY;
const allInputTypeBitmask: BitmaskFlags =
    BitmaskFlags.NONE_INPUT | BitmaskFlags.CANCEL | BitmaskFlags.DOWN | BitmaskFlags.MOVE | BitmaskFlags.UP;
const allInteractionTypeBitmask: BitmaskFlags =
    BitmaskFlags.GRAB | BitmaskFlags.HOVER | BitmaskFlags.PUSH | BitmaskFlags.TOUCHPLANE | BitmaskFlags.VELOCITYSWIPE;
const bitmaskFlagParams: BitmaskFlags[] = [
    // The parameters here are hand-written and non-exhaustive as the cost
    // of exhaustive combinations in maintenance and performance is excessive

    // All set
    allHandChiralityBitmask | allHandTypeBitmask | allInputTypeBitmask | allInteractionTypeBitmask,
    // Individual values
    BitmaskFlags.LEFT,
    BitmaskFlags.RIGHT,

    BitmaskFlags.PRIMARY,
    BitmaskFlags.SECONDARY,

    BitmaskFlags.NONE_INPUT,
    BitmaskFlags.CANCEL,
    BitmaskFlags.DOWN,
    BitmaskFlags.MOVE,
    BitmaskFlags.UP,

    BitmaskFlags.GRAB,
    BitmaskFlags.HOVER,
    BitmaskFlags.PUSH,
    BitmaskFlags.TOUCHPLANE,
    BitmaskFlags.VELOCITYSWIPE,
    // One of each enum
    BitmaskFlags.LEFT | BitmaskFlags.PRIMARY | BitmaskFlags.DOWN | BitmaskFlags.GRAB,
    BitmaskFlags.RIGHT | BitmaskFlags.SECONDARY | BitmaskFlags.NONE_INPUT | BitmaskFlags.TOUCHPLANE,
    BitmaskFlags.LEFT | BitmaskFlags.SECONDARY | BitmaskFlags.UP | BitmaskFlags.VELOCITYSWIPE,
    BitmaskFlags.RIGHT | BitmaskFlags.PRIMARY | BitmaskFlags.HOVER | BitmaskFlags.PUSH,
    // Multiple of same enum set
    allHandChiralityBitmask,
    allHandTypeBitmask,
    allInputTypeBitmask,
    allInteractionTypeBitmask,
    // Nothing set
    BitmaskFlags.NONE,
];

describe('BitmaskFlag', () => {
    // Suppress errors from console and store them in an array which we print only if a test fails
    const errors: string[] = [];
    const consoleErrorMock = jest.spyOn(console, 'error').mockImplementation((msg: string) => errors.push(msg));

    it('should combine the same as before', () => {
        expect(FlagUtilities.GetInteractionFlags).toVerifyAllCombinations(
            interactionTypes,
            handTypes,
            handChiralities,
            inputTypes
        );
    });

    it('should deserialize hand chirality the same as before', () => {
        expect(FlagUtilities.GetChiralityFromFlags).toVerifyAllCombinations(bitmaskFlagParams);
    });

    it('should deserialize hand type the same as before', () => {
        expect(FlagUtilities.GetHandTypeFromFlags).toVerifyAllCombinations(bitmaskFlagParams);
    });

    it('should deserialize input type the same as before', () => {
        expect(FlagUtilities.GetInputTypeFromFlags).toVerifyAllCombinations(bitmaskFlagParams);
    });

    it('should deserialize interaction type the same as before', () => {
        expect(FlagUtilities.GetInteractionTypeFromFlags).toVerifyAllCombinations(bitmaskFlagParams);
    });

    afterAll(() => {
        consoleErrorMock.mockRestore();
        // If not all tests pass then log the errors
        const JEST_MATCHER = Symbol.for('$$jest-matchers-object');
        // eslint-disable-next-line @typescript-eslint/no-explicit-any
        const numberOfNonMatches: number | undefined = (global as any)[JEST_MATCHER]?.state?.snapshotState?.unmatched;
        if (numberOfNonMatches && numberOfNonMatches > 0) {
            console.log(errors);
        }
    });
});
