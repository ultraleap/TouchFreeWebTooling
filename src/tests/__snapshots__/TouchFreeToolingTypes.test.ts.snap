// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BitmaskFlag should combine the same as before 1`] = `
{
  "0,0,0,0": 533,
  "0,0,0,1": 549,
  "0,0,0,2": 581,
  "0,0,0,3": 645,
  "0,0,0,4": 773,
  "0,0,1,0": 534,
  "0,0,1,1": 550,
  "0,0,1,2": 582,
  "0,0,1,3": 646,
  "0,0,1,4": 774,
  "0,1,0,0": 537,
  "0,1,0,1": 553,
  "0,1,0,2": 585,
  "0,1,0,3": 649,
  "0,1,0,4": 777,
  "0,1,1,0": 538,
  "0,1,1,1": 554,
  "0,1,1,2": 586,
  "0,1,1,3": 650,
  "0,1,1,4": 778,
  "1,0,0,0": 1045,
  "1,0,0,1": 1061,
  "1,0,0,2": 1093,
  "1,0,0,3": 1157,
  "1,0,0,4": 1285,
  "1,0,1,0": 1046,
  "1,0,1,1": 1062,
  "1,0,1,2": 1094,
  "1,0,1,3": 1158,
  "1,0,1,4": 1286,
  "1,1,0,0": 1049,
  "1,1,0,1": 1065,
  "1,1,0,2": 1097,
  "1,1,0,3": 1161,
  "1,1,0,4": 1289,
  "1,1,1,0": 1050,
  "1,1,1,1": 1066,
  "1,1,1,2": 1098,
  "1,1,1,3": 1162,
  "1,1,1,4": 1290,
  "2,0,0,0": 2069,
  "2,0,0,1": 2085,
  "2,0,0,2": 2117,
  "2,0,0,3": 2181,
  "2,0,0,4": 2309,
  "2,0,1,0": 2070,
  "2,0,1,1": 2086,
  "2,0,1,2": 2118,
  "2,0,1,3": 2182,
  "2,0,1,4": 2310,
  "2,1,0,0": 2073,
  "2,1,0,1": 2089,
  "2,1,0,2": 2121,
  "2,1,0,3": 2185,
  "2,1,0,4": 2313,
  "2,1,1,0": 2074,
  "2,1,1,1": 2090,
  "2,1,1,2": 2122,
  "2,1,1,3": 2186,
  "2,1,1,4": 2314,
  "3,0,0,0": 4117,
  "3,0,0,1": 4133,
  "3,0,0,2": 4165,
  "3,0,0,3": 4229,
  "3,0,0,4": 4357,
  "3,0,1,0": 4118,
  "3,0,1,1": 4134,
  "3,0,1,2": 4166,
  "3,0,1,3": 4230,
  "3,0,1,4": 4358,
  "3,1,0,0": 4121,
  "3,1,0,1": 4137,
  "3,1,0,2": 4169,
  "3,1,0,3": 4233,
  "3,1,0,4": 4361,
  "3,1,1,0": 4122,
  "3,1,1,1": 4138,
  "3,1,1,2": 4170,
  "3,1,1,3": 4234,
  "3,1,1,4": 4362,
  "4,0,0,0": 8213,
  "4,0,0,1": 8229,
  "4,0,0,2": 8261,
  "4,0,0,3": 8325,
  "4,0,0,4": 8453,
  "4,0,1,0": 8214,
  "4,0,1,1": 8230,
  "4,0,1,2": 8262,
  "4,0,1,3": 8326,
  "4,0,1,4": 8454,
  "4,1,0,0": 8217,
  "4,1,0,1": 8233,
  "4,1,0,2": 8265,
  "4,1,0,3": 8329,
  "4,1,0,4": 8457,
  "4,1,1,0": 8218,
  "4,1,1,1": 8234,
  "4,1,1,2": 8266,
  "4,1,1,3": 8330,
  "4,1,1,4": 8458,
}
`;

exports[`BitmaskFlag should deserialize hand chirality the same as before 1`] = `
{
  "0": 1,
  "1": 0,
  "1024": 1,
  "12": 1,
  "128": 1,
  "15872": 1,
  "16": 1,
  "16383": 1,
  "2": 1,
  "2048": 1,
  "256": 1,
  "3": 1,
  "3078": 1,
  "32": 1,
  "4": 1,
  "4096": 1,
  "4122": 1,
  "496": 1,
  "512": 1,
  "581": 0,
  "64": 1,
  "8": 1,
  "8192": 1,
  "8457": 0,
}
`;

exports[`BitmaskFlag should deserialize hand type the same as before 1`] = `
{
  "0": 0,
  "1": 0,
  "1024": 0,
  "12": 0,
  "128": 0,
  "15872": 0,
  "16": 0,
  "16383": 0,
  "2": 0,
  "2048": 0,
  "256": 0,
  "3": 0,
  "3078": 0,
  "32": 0,
  "4": 0,
  "4096": 0,
  "4122": 1,
  "496": 0,
  "512": 0,
  "581": 0,
  "64": 0,
  "8": 1,
  "8192": 0,
  "8457": 1,
}
`;

exports[`BitmaskFlag should deserialize input type the same as before 1`] = `
{
  "0": 0,
  "1": 0,
  "1024": 0,
  "12": 0,
  "128": 3,
  "15872": 0,
  "16": 0,
  "16383": 0,
  "2": 0,
  "2048": 0,
  "256": 4,
  "3": 0,
  "3078": 0,
  "32": 1,
  "4": 0,
  "4096": 0,
  "4122": 0,
  "496": 0,
  "512": 0,
  "581": 2,
  "64": 2,
  "8": 0,
  "8192": 0,
  "8457": 4,
}
`;

exports[`BitmaskFlag should deserialize interaction type the same as before 1`] = `
{
  "0": 2,
  "1": 2,
  "1024": 1,
  "12": 2,
  "128": 2,
  "15872": 2,
  "16": 2,
  "16383": 2,
  "2": 2,
  "2048": 2,
  "256": 2,
  "3": 2,
  "3078": 2,
  "32": 2,
  "4": 2,
  "4096": 3,
  "4122": 3,
  "496": 2,
  "512": 0,
  "581": 0,
  "64": 2,
  "8": 2,
  "8192": 4,
  "8457": 4,
}
`;
